SinkManyUnicast discard support during subscription cancel

chemicL opened the pull request:
When subscription to `SinkManyUnicast` is cancelled at the time of being delivered, the discarding fails, as the actual `Subscriber` is not yet assigned. This change reorders the operations so that the actual is stored properly before delivering the `Subscription`.

chemicL commented:
Analysing the test failures.

OlegDokuka reviewed (changes requested):
This may cause issues of concurrent onNext with onSubscribe or even onNext hapenning before onSubscribe

chemicL commented on reactor-core/src/test/java/reactor/test/publisher/BaseOperatorTest.java:
```diff
@@ -595,20 +595,14 @@ final void inputFusedAsyncOutputFusedAsyncCancel(OperatorScenario<I, PI, O, PO>
-						            assertThat(qs.size()).isEqualTo(up.scan(Attr.BUFFERED)); //UnicastProcessor#size
-					            }
 					            try {
+									// Assuming scenario.producing >= 3
```
This seems to have been a mistake as on line 598 the demand was 0. So it was actually illegal to poll, which causes onNext signals to be triggered. In cases such as filtering out all the values, this would trigger upstream demand to continuously increase forcing termination. Therefore, the initial demand is set to scenario.producing and this code assumes it's not less than 3 to meet the three polls that happen here.

chemicL replied:
Actually, as discussed with @OlegDokuka `poll`-ing expresses demand on its own, so can be an optimization technique in these fused scenarios and not really a misuse. The issue observed was with not protecting against simultaneous `onNext` that can happen as a result of calling `poll` during `onSubscribe`. Now with proper protection in place, the problem does not occur.

chemicL commented:
> This may cause issues of concurrent onNext with onSubscribe or even onNext hapenning before onSubscribe

Thanks, I updated the code to address this as discussed and added a few notes in this PR to clarify the changes.
